### DISCUSSION TITLE
Forced upgrades of Python version no thanks

### DISCUSSION THREAD

**Discussion by mok0** [Homebrew/discussions#250] [Everyday usage]
Whenever I do a `brew upgrade` homebrew currently installs Python 3.9 as the default `/usr/local/bin/python3` but since I am using lots of modules from PyPi that are not upgraded (and indeed some aren't properly ported to 3.9 yet) I don't want that. I am in the middle of a project and I am not ready to upgrade Python.

I think it is improper of homebrew to force an upgrade of the Python version upon you. The right way to do this is to add a Python environment management system that lets you choose which version of Python is your main one, for example `pyenv` or something similar. It could also be a brew command. This should not be done automatically at upgrade.

**Comment by SMillerDev** [comment]
> The right way to do this is to add a Python environment management system that lets you choose which version of Python is your main one, for example pyenv or something similar. It could also be a brew command. This should not be done automatically at upgrade.

So... `brew install pyenv`? You can already run that and use `pyenv` to manage your install. Brew will still update your software when you do a `brew upgrade` though, that is literally what you're asking it to do.

**Reply by mok0** [reply]
`pyenv` installs Python from source code and literally defeats the purpose of Homebrew.

**Reply by SMillerDev** [reply]
It doesn't install anything from source and it works alongside Homebrew. If you need to have control over your python version: this is the way to go, not requesting that Homebrew doesn't update Python until you're done with the older version. That's simply not how package managers work.

**Reply by mok0** [reply]
Yes it does, it downloads a tarball of the Python sources compiles it and installs it under ~/.pyenv. Granted, the compilation will use any homebrew installed libraries, like readline.

```
homunculus➜  ~  pyenv install 3.8.6
python-build: use openssl@1.1 from homebrew
python-build: use readline from homebrew
Downloading Python-3.8.6.tar.xz...
-> https://www.python.org/ftp/python/3.8.6/Python-3.8.6.tar.xz
Installing Python-3.8.6...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
^C

BUILD FAILED (OS X 10.15.7 using python-build 20180424)
```

If it is possible to make it manage homebrew binaries please tell us how it is done.

Edit: I am not looking to pin an older version of Python. It's simply about which binary the symlink `/usr/local/bin/python3`, `/usr/local/bin/pip3` and a few others point to.

**Reply by SMillerDev** [reply]
> If it is possible to make it manage homebrew binaries please tell us how it is done.

As far as I know it can use the homebrew version if available.

> Edit: I am not looking to pin an older version of Python. It's simply about which binary the symlink /usr/local/bin/python3, /usr/local/bin/pip3 and a few others point to.

`brew unlink python@3.9`  and `brew link python@3.8`

**Comment by mok0** [comment]
Thanks, those commands will spare me from doing it manually! :-)  And I maintain `pyenv` cannot use the Homebrew binaries, it's not how it works. It has a huge list of Python versions it knows about, and it builds them all from source. You can see this list using `pyenv install --list`. If effect, `pyenv` is Homebrew for Python.

I also maintain it is improper for Homebrew to upgrade Python versions other than microversions behind your back: When I have done `brew upgrade`, and some time later come to invoke `python3` I no longer have access to the tons of PyPi modules I am using. And for example, `scikit-image` cannot compile under Python 3.9. So I am not ready to move to Python 3.9 until all the modules are working.

Linux distros don't do it that way either. Debian & derivatives use the `update-alternatives` system, which is a system of symlinks that lets the sysadm decide which particular version the generic program names like `python`, `editor` or `awk` point to. 

I hope the Homebrew crew will be able to work out a better way to manage this.

**Reply by iMichka** [reply]
You can now `brew install python@3.9`, `brew install python@3.8`, `brew install python@3.7`.

When we add python 3.10 you will be able to work with `brew install python@3.10`.

Have a look at the `brew info python@3.x` to check for the needed paths. We recommend that you use these paths as they will not change over time and will be more stable.

We just switch the main python on each major release: if you need stability, do not use the /opt path, use the absolute paths to the specific version you need. You can also create aliases in your .bashrc/.zshrc if needed.

**Reply by mok0** [reply]
Thanks a lot, I think this info can help solve my problem!

**Reply by iMichka** [reply]
This is quite new, since the big python 3.7 to 3.8 migration. Before we would do the switch and it would break people's stuff. This is why now each python formula is sort of self-contained. If you work with the absolute paths you should be much safer.

**Reply by mok0** [reply]
I am not sure how Python 3.9 (python@3.9) appeared on my machine though. I'm sure I didn't install it manually.

**Reply by mok0** [reply]
Btw, Homebrew needs the same system for `ipython`, e.g. `ipython@3.8` etc.

**Reply by iMichka** [reply]
> I am not sure how Python 3.9 (python@3.9) appeared on my machine though. I'm sure I didn't install it manually.

A lot of formulae depend on the latest python. You can check this with `brew uses python@3.9`. As soon as you installed one of these and `brew upgrade`, the latest python will be installed on your system by brew.

> Btw, Homebrew needs the same system for ipython, e.g. ipython@3.8 etc.

Pull requests are welcome to add it. You are the first one to ask for this, so not sure we will add it (it depends on how many people want/need it). But you still host that formula in a personal tap.
